File: ServiceCoordinationService/src/middleware/authMiddleware.js
import jwt from "jsonwebtoken";

export const protect = (allowedRoles = []) => {
  return (req, res, next) => {
    try {
      const authHeader = req.headers.authorization;

      if (!authHeader || !authHeader.startsWith("Bearer ")) {
        return res.status(401).json({
          success: false,
          message: "Authorization token is required",
        });
      }

      const token = authHeader.split(" ")[1];
      const decoded = jwt.verify(token, process.env.JWT_SECRET);

      // Support nested token payload: { user: { id, role } }
      const userId =
        decoded.user?.id ||
        decoded.user?._id ||
        decoded.id ||
        decoded._id ||
        decoded.userId ||
        decoded.providerId ||
        decoded.seekerId ||
        decoded.adminId ||
        decoded.sub;

      const role =
        decoded.user?.role ||
        decoded.role ||
        decoded.userRole ||
        decoded.type ||
        decoded.accountType;

      if (!userId || !role) {
        return res.status(401).json({
          success: false,
          message: "Invalid token payload",
        });
      }

      req.user = {
        id: userId,
        role,
        email: decoded.user?.email || decoded.email || "",
      };

      if (allowedRoles.length > 0 && !allowedRoles.includes(req.user.role)) {
        return res.status(403).json({
          success: false,
          message: "Access denied for this role",
          userRole: req.user.role,
          allowedRoles,
        });
      }

      next();
    } catch (error) {
      return res.status(401).json({
        success: false,
        message: "Invalid or expired token",
        error: error.message,
      });
    }
  };
};

File: Provider_Service/ML_Engine/src/middleware/auth_middleware.py
import os
from functools import wraps
import jwt
from flask import request, jsonify

JWT_SECRET    = os.getenv("JWT_SECRET", "")
JWT_ALGORITHM = os.getenv("JWT_ALGORITHM", "HS256")

def jwt_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        if not auth_header.startswith("Bearer "):
            return jsonify({"error": "Authorization header missing."}), 401
        token = auth_header.split(" ", 1)[1].strip()
        try:
            payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token expired."}), 401
        except jwt.InvalidTokenError as e:
            return jsonify({"error": f"Invalid token: {e}"}), 401
        user_obj = payload.get("user", {})
        user_id  = user_obj.get("id") or user_obj.get("_id")
        if not user_id:
            return jsonify({"error": "Token payload missing user ID."}), 401
        return f(str(user_id), *args, **kwargs)
    return decorated